File: golang/0238.go
func productExceptSelf(nums []int) []int {
	length := len(nums)
	left := make([]int, length)
	right := make([]int, length)
	res := make([]int, length)

	left[0] = 1
	right[length-1] = 1

	for i := 1; i < length; i++ {
		left[i] = left[i-1] * nums[i-1]
	}

	for i := length - 2; i >= 0; i-- {
		right[i] = right[i+1] * nums[i+1]
	}

	for i := 0; i < length; i++ {
		res[i] = left[i] * right[i]
	}

	return res
}

File: rust/0238.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let mut res: Vec<i32> = vec![0; nums.len()];
        let mut left = vec![0; nums.len()];
        let mut right = vec![0; nums.len()];

        left[0] = 1;
        for i in 1..nums.len() {
            left[i] = left[i - 1] * nums[i - 1]
        }

        right[nums.len() - 1] = 1;
        for i in (0..nums.len()-1).rev() {
            right[i] = right[i + 1] * nums[i + 1];
        }

        for i in 0..nums.len() {
            res[i] = left[i] * right[i]
        }

        res
    }
}
